apps/dashboard/src/components/primitives/field-editor/field-editor.tsx

```diff
@@ -0,0 +1,232 @@
+  const isUpdatingRef = useRef(false);
+
+  const handleVariableSelect = useCallback((value: string, from: number, to: number) => {
+    if (isUpdatingRef.current) return;
+
+    requestAnimationFrame(() => {
+      setSelectedVariable({ value, from, to });
+    });
+  }, []);
+
+  return {
+    selectedVariable,
+    setSelectedVariable,
+    handleVariableSelect,
+    isUpdatingRef,
+  };
+}
+
+/**
+ * Handles the logic of updating Liquid variables in the editor while maintaining proper syntax.
+ *
+ * This hook manages several critical aspects of variable editing:
+ * 1. Ensures proper Liquid syntax ({{...}}) is maintained when updating variables
+ * 2. Handles edge cases like existing closing brackets
+ * 3. Maintains cursor position after updates
+ * 4. Prevents recursive updates during the edit process
+ *
+ * The update process:
+ * 1. Checks if the new value already has Liquid syntax
+ * 2. Detects if there are existing closing brackets after the cursor
+ * 3. Calculates the correct range to replace
+ * 4. Updates the editor state while maintaining proper cursor position
+ */
+function useVariableUpdate(
+  selectedVariable: SelectedVariable,
+  viewRef: React.RefObject<EditorView>,
+  isUpdatingRef: React.MutableRefObject<boolean>,
+  onChange: (value: string) => void,
+  setSelectedVariable: Dispatch<SetStateAction<SelectedVariable>>
+) {
+  return useCallback(
+    (newValue: string) => {
+      if (!selectedVariable || !viewRef.current || isUpdatingRef.current) return;
+
+      try {
+        isUpdatingRef.current = true;
+        const { from, to } = selectedVariable;
+        const view = viewRef.current;
+
+        const hasLiquidSyntax = newValue.match(/^\{\{.*\}\}$/);
+        const newVariableText = hasLiquidSyntax ? newValue : `{{${newValue}}}`;
+
+        const currentContent = view.state.doc.toString();
+        const afterCursor = currentContent.slice(to).trim();
+        const hasClosingBrackets = afterCursor.startsWith('}}');
+
+        const changes = {
+          from,
+          to: hasClosingBrackets ? to + 2 : to,
+          insert: newVariableText,
+        };
+
+        view.dispatch({
+          changes,
+          selection: { anchor: from + newVariableText.length },
+        });
+
+        onChange(view.state.doc.toString());
+
+        setSelectedVariable((prev: SelectedVariable) =>
+          prev ? { ...prev, value: newValue, to: from + newVariableText.length } : null
+        );
+      } finally {
+        isUpdatingRef.current = false;
+      }
+    },
+    [selectedVariable, onChange, viewRef, isUpdatingRef, setSelectedVariable]
+  );
+}
+
+function useCompletionSource(variables: LiquidVariable[], lastCompletionRef: React.MutableRefObject<CompletionRange>) {
+  return useCallback(
+    (context: CompletionContext) => {
+      // Match text that starts with {{ and capture everything after it until the cursor position
+      const word = context.matchBefore(/\{\{([^}]*)/);
```

Comment:
We should move all the parsing logic into the completions utility and simplify the hook.